### ISSUE TITLE
CJellyfish.soundex() chokes on bad Unicode string which Jellyfish.soundex() can handle

### ISSUE THREAD

**Issue by nchammas** [jamesturk/jellyfish#63]
Data is messy and oftentimes we get strings that have been encoded incorrectly. Detecting the correct encoding and fixing the mal-encoded strings seems difficult to automate.

In this particular case, the C implementation of `soundex()` seems to choke on a bad Unicode string that the Python implementation can handle fine. It would be ideal if the C implementation behaved like in the Python one.

Repro in Python 3.5:

```python
>>> s = b'\xef\xbf\xbd1 Championship Dr'.decode('utf-8')
>>> import jellyfish
>>> jellyfish.soundex(s)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xff in position 0: invalid start byte
>>> import jellyfish._jellyfish as jellyfish
>>> jellyfish.soundex(s)
'�251'
```

Relates to #55 and #56 in that the C implementations of certain functions don't handle Unicode (whether good or bad) the same way that the Python implementations do.

**Comment by nchammas** [comment]
Another example:

```python
>>> import jellyfish
>>> jellyfish.soundex('¿')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xff in position 0: invalid start byte
>>> import jellyfish._jellyfish as pyjellyfish
>>> pyjellyfish.soundex('¿')
'¿000'
```

**Comment by jsfenfen** [comment]
@nchammas: are you saying the correct behavior is for jellyfish to choke on the unicode or not? "Soundex is a phonetic algorithm for indexing names by sound, as pronounced in English" is how wikipedia puts it. Personally I'd expect it to choke on things that can't be "pronounced in English". If there's a reference implementation of this in other languages I'd be interested in seeing it, but my understanding is that this is a system designed circa 1920 around how something sounds when said out loud in English.  If we knew how people said ¿ in 1919 we might be able to implement it, but...

**Comment by nchammas** [comment]
I'm not sure what the correct behavior is, honestly. I can see the case for throwing an exception, and I can see the case for returning the input as-is or something similar.  

What I _am_ sure of, however, is that it's not correct for the C and Python implementations of the same function to behave differently when given the same input.

**Comment by J535D165** [comment]
Also, the Jellyfish Soundex algorithm handles whitespace as a letter. 

```
In [1]: jellyfish.soundex(' Jellyfish')
Out[1]: ' 241'

In [2]: jellyfish.soundex('Jellyfish')
Out[2]: 'J412'
```

When the first character is a number:

```
In [3]: jellyfish.soundex('2Pac')
Out[3]: '2120'
```

I like [this](http://creativyst.com/Doc/Articles/SoundEx1/SoundEx1.htm) implementation of the Soundex algorithm. It solves all issues mentioned in this topic.

**Comment by jamesturk** [comment]
fixed via #177
